packages/test-utils/src/matches.js

```diff
@@ -4,10 +4,20 @@ import {
   FUNCTIONAL_OPTIONS
 } from 'shared/consts'
 import { isConstructor } from 'shared/validators'
+import { capitalize, camelize } from 'shared/util'
 
-export function vmMatchesName(vm, name) {
+function vmMatchesName(vm, name) {
+  console.log(name)
+  console.log(vm.$options.name)
   return (
-    !!name && (vm.name === name || (vm.$options && vm.$options.name === name))
+    !!name && (
+      vm.name === name ||
+      (vm.$options && vm.$options.name === name) ||
+      vm.name === capitalize(name) ||
+      vm.$options && vm.$options.name === capitalize(name) ||
+      vm.$options && vm.$options.name && vm.$options.name === capitalize(camelize(name)) ||
+      vm.$options && vm.$options.name && capitalize(camelize(vm.$options.name)) === name
+    )
```

Comment:
I have a proposal, lets save the vm.$options.name to a variable, something like 
```js
const componentName = vm.$options && vm.$options.name || ''
```
then we can easily reduce those those lines. Yes it will use a bit more memory, but for readability sake I think it makes sense?
Also that line 18 has an extra check others dont.
Unrelated:
You know whats the diff between `vm.name` and `vm.$options.name`? One is the one you set in its options and the other is the one that it is registered by?

Reply:
Can we add comments for each of the conditionals? Eg what use case each line is handling? A 7 line if statement is hard to grok easily.

Reply:
Yep agreed, I'll make these changes

Reply:
@dobromir-hristov I am not sure on the difference between `vm.name` and `vm.$options.name`. `vm.name` is undefined in every test... I removed it and everything is still passing. If it serves no purpose, we may as well omit that check. What do you think?

We should check the source code for Vue and see if we can figure out the difference.

Regarding your comment about the extra variable - I agree, that is more readable. I updated it the code. 

@JessicaSachs I added some comments explaining why we do each of the matches. What do you think? If you think you can make the comment clearer, you can add some commits or I can add some if you leave an updated comment here.

Reply:
I also gave a quick search and did not see `vm.name` defined. Its not in the Vue types as well. 
Wonder if its something from "old days"? 
New code is much more readable, gj